v2/googlecloud-to-neo4j/src/main/java/com/google/cloud/teleport/v2/neo4j/GcpToNeo4j.java

```diff
@@ -0,0 +1,339 @@
+     */
+    public GcpToNeo4j(final Neo4jFlexTemplateOptions pipelineOptions) {
+        ////////////////////////////
+        // We need to initialize pipeline in order to create context for Gs and Bq file system
+        final String jobName = pipelineOptions.getJobName() + "-" + System.currentTimeMillis();
+        pipelineOptions.setJobName(jobName);
+
+        // Set pipeline options
+        this.pipeline = Pipeline.create(pipelineOptions);
+        FileSystems.setDefaultPipelineOptions(pipelineOptions);
+        this.optionsParams = OptionsParamsMapper.fromPipelineOptions(pipelineOptions);
+        // Validate pipeline
+        processValidations(InputValidator.validateNeo4jPipelineOptions(pipelineOptions));
+
+        this.neo4jConnection = new ConnectionParams(pipelineOptions.getNeo4jConnectionUri());
+        // Validate connection
+        processValidations(InputValidator.validateNeo4jConnection(this.neo4jConnection));
+
+        this.jobSpec = JobSpecMapper.fromUri(pipelineOptions.getJobSpecUri());
+        // Validate job spec
+        processValidations(InputValidator.validateJobSpec(this.jobSpec));
+
+        ///////////////////////////////////
+        // Refactor job spec
+        InputRefactoring inputRefactoring = new InputRefactoring(this.optionsParams);
+        // Variable substitution
+        inputRefactoring.refactorJobSpec(this.jobSpec);
+        // Optimizations
+        inputRefactoring.optimizeJobSpec(this.jobSpec);
+
+        ///////////////////////////////////
+        // Text input specific options and validation
+        if (this.jobSpec.sources.size() == 0) {
+            String errMsg = "JobSpec source is required.";
+            LOG.error(errMsg);
+            throw new RuntimeException(errMsg);
+        }
+        // Source specific validations
+        for (Source source : jobSpec.getSourceList()) {
+            //get provider implementation for source
+            IProvider providerImpl = ProviderFactory.of(source.sourceType);
+            providerImpl.configure(optionsParams, jobSpec);
+            processValidations(providerImpl.validateJobSpec());
+        }
+
+        //LOG.info(gson.toJson(jobSpec));
+    }
+
+    /**
+     * Runs a pipeline which reads data from various sources and writes it to Neo4j.
+     *
+     * @param args arguments to the pipeline
+     */
+    public static void main(final String[] args) {
+        final Neo4jFlexTemplateOptions options =
+                PipelineOptionsFactory.fromArgs(args).withValidation()
+                        .as(Neo4jFlexTemplateOptions.class);
+
+        LOG.info("Job: " + options.getJobSpecUri());
+        final GcpToNeo4j bqToNeo4jTemplate = new GcpToNeo4j(options);
+        bqToNeo4jTemplate.run();
+    }
+
+    /**
+     * Raises RuntimeExceptions for validation errors.
+     *
+     * @param validationMessages
+     */
+    private void processValidations(List<String> validationMessages) {
+        StringBuffer sb = new StringBuffer();
```

Comment:
A few issues with this block:

1. If you keep StringBuffer, then replace it with StringBuilder. StringBuffer uses excessive synchronization, which is not needed here.
2. Logging and throwing the same exception is usually an antipattern. In this case we'll end up with the same messages logged twice in the Dataflow job logs. I would either remove logging and throw the exception only, or remove validationMessages from the exception message. Log the validationMessages and then throw "Errors found validating pipeline input. Please see logs for more details." exception, without any details.
3. If you remove logging and keep the exception, I think you also need to add line breaks or another separator between the validation messages to StringBuilder.

Reply:
Removed the logging when we're sending validation errors back to the UI.